executor/adapter.go

```diff
@@ -141,8 +142,40 @@ func (a *statementAdapter) Exec(ctx context.Context) (rset.Recordset, error) {
 	if b.err != nil {
 		return nil, errors.Trace(b.err)
 	}
-	fields := make([]*field.ResultField, 0, len(e.Fields()))
-	for _, v := range e.Fields() {
+
+	if executorExec, ok := e.(*ExecuteExec); ok {
+		err := executorExec.Build()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if executorExec.OldStmt != nil {
+			return executorExec.OldStmt.Exec(executorExec.Ctx)
+		}
+		e = executorExec.StmtExec
+	}
+
+	if len(e.Fields()) == 0 {
+		// No result fields means no Recordset.
+		defer e.Close()
+		for {
+			row, err := e.Next()
```

Comment:
For `Update` and `Delete` statement, we call `Next` in for loop, so we can separate the task into many steps, so we can cancel the task in the middle.